tracer.go

```diff
@@ -121,6 +127,118 @@ type StartSpanOptions struct {
 	// identical to those for Span.SetTag(). May be nil.
 	//
 	// If specified, the caller hands off ownership of Tags at
-	// StartSpanWithOptions() invocation time.
+	// StartSpan() invocation time.
 	Tags map[string]interface{}
 }
+
+// StartSpanOption instances (zero or more) may be passed to Tracer.StartSpan.
+type StartSpanOption interface {
+	Apply(*StartSpanOptions)
+}
+
+// ReferenceType is an enum type describing different categories of
+// relationships between two Spans. If Span-2 refers to Span-1, the
+// ReferenceType describes Span-1 from Span-2's perspective. For example,
+// BlockedParent means that Span-1 caused Span-2, and that it's blocked on
+// Span-2's finish.
+//
+// NOTE: Span-1 and Span-2 do *not* necessarily depend on each other for
+// completion; e.g., Span-2 may be part of a background job enqueued by Span-1,
+// or Span-2 may be sitting in a distributed queue behind Span-1.
+type ReferenceType int
+
+const (
+	// BlockedParent refers to a "parent" Span that (a) caused the creation of
+	// the newly-started span, and (b) is blocked on the newly-started Span.
+	// See StartedBefore for the non-blocking version. Timing diagram:
+	//
+	//     [-Referent------------]
+	//               [-New Span-]
+	//
+	BlockedParent ReferenceType = iota
+
+	// RPCClient is a special case of BlockedParent for the server side of RPCs
+	// to refer to the client side of RPCs. Timing diagram:
+	//
+	//     [-Remote Client Referent------]
+	//                [-New Server Span-]
+	//
+	// This is a special case as some tracing implementations (e.g., Dapper and
+	// Zipkin) consider the two sides of an RPC to be a single Span and thus
+	// need to distinguish this scenario at the OpenTracing layer.
+	RPCClient
+
+	// StartedBefore refers to a Span that merely started before the
+	// newly-started Span. Timing diagram:
+	//
+	//     [-Referent-]
+	//          [-New Span-]
+	//
+	// For example, New Span might be a best-effort, fire-and-forget
+	// distributed cache insert.
+	StartedBefore
+
+	// FinishedBefore refers to a Span that finished (and started) before
+	// the newly-created Span. Timing diagram:
+	//
+	//     [-Referent-]
+	//                    [-New Span-]
+	//
+	// For example, New Span might be stage N+1 of a batch processing pipeline
+	// (where Referent represents stage N of same).
+	FinishedBefore
+)
+
+// SpanReference pairs a ReferenceType and a referent SpanMetadata. See the
+// ReferenceType documentation.
+type SpanReference struct {
+	Type     ReferenceType
+	Metadata SpanMetadata
+}
+
+// Apply satisfies the StartSpanOption interface.
+func (r SpanReference) Apply(o *StartSpanOptions) {
+	o.References = append(o.References, r)
+}
+
+// IntraTrace returns true if the referent is in the same trace as the referee
+// (i.e., if the referee should adopt the referent's trace-scoped metadata,
+// like a `trace_id`).
+func (r ReferenceType) IntraTrace() bool {
+	return r <= FinishedBefore
```

Comment:
Upon reflection, I think it would be better to get rid of `IntraTrace()` entirely. It's not needed for this PR and IMO it's too difficult to resolve threads like this one without more concrete use cases.